### PR TITLE
Add Kahn's algorithm to graph.h

### PR DESCRIPTION
**Pull request by mgudemann**
This adds topological sorting to `graph.h`.

### CODE REVIEW

File: src/util/graph.h [hunk 1]
```diff
@@ -19,6 +19,9 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <ostream>
 #include <cassert>
 #include <algorithm>
+#include <queue>
+
+#include <util/invariant.h>
```
Comment (tautschnig) [inline src/util/graph.h]: Nit picking: in `util` we just use `#include "invariant.h"`

File: src/util/graph.h [hunk 2]
```diff
@@ -570,6 +579,54 @@ void grapht<N>::make_chordal()
...
+      // remove edge from graph, by decrementing the in-degree of target
+      // outgoing edges from source will not be traversed again
+      in_deg[target]--;
+      if(in_deg[target]==0)
+        indeg0_nodes.push(target);
+    }
+  }
+
+  // if no edge is left in the graph, it is acyclic
+  bool is_dag=none_of(
```
Comment (tautschnig) [inline src/util/graph.h]: Doesn't this need a `std::` ?

File: src/util/graph.h [hunk 3]
```diff
@@ -226,6 +229,12 @@ class grapht
   // return value: number of SCCs
   std::size_t SCCs(std::vector<node_indext> &subgraph_nr);
 
+  bool is_dag()
```
Comment (tautschnig) [inline src/util/graph.h]: Should be marked `const`

File: src/util/graph.h [hunk 4]
```diff
@@ -226,6 +229,12 @@ class grapht
...
+  {
+    return size()==0 || !topsort().empty();
+  }
+  std::list<node_indext> topsort();
```
Comment (tautschnig) [inline src/util/graph.h]: Should be marked `const` (plus add a blank line in between)

File: src/util/graph.h [hunk 5]
```diff
@@ -226,6 +229,12 @@ class grapht
   // return value: number of SCCs
   std::size_t SCCs(std::vector<node_indext> &subgraph_nr);
 
+  bool is_dag()
+  {
+    return size()==0 || !topsort().empty();
```
Comment (tautschnig) [inline src/util/graph.h]: I believe we've now got `empty()`

File: src/util/graph.h [hunk 6]
```diff
@@ -570,6 +579,54 @@ void grapht<N>::make_chordal()
...
 }
 
+/// Find a topological order of the nodes if graph is DAG, return empty list for
+/// non-DAG or empty graph. Uses Kahn's algorithm running in O(#edges+#nodes).
+template<class N>
+std::list<typename grapht<N>::node_indext> grapht<N>::topsort()
+{
+  std::list<node_indext> nodelist;
+  std::queue<node_indext> indeg0_nodes;
+  std::map<node_indext, size_t> in_deg;
```
Comment (tautschnig) [inline src/util/graph.h]: I would recommend using a zero-initialised `std::vector` of size `nodes.size()`

File: src/util/graph.h [hunk 7]
```diff
@@ -570,6 +579,54 @@ void grapht<N>::make_chordal()
...
+  while(!indeg0_nodes.empty())
+  {
+    node_indext source=indeg0_nodes.front();
+    indeg0_nodes.pop();
+    nodelist.push_back(source);
+
+    for(const auto &edge : out(source))
+    {
+      const node_indext target=edge.first;
+      INVARIANT(in_deg[target]!=0, "In-degree of node cannot be 0.");
```
Comment (tautschnig) [inline src/util/graph.h]: Nit picking/question: with `throw` we even have a lint check that the message does not start with an uppercase character; should the same apply to invariants?

File: src/util/graph.h [hunk 8]
```diff
@@ -570,6 +579,54 @@ void grapht<N>::make_chordal()
...
+      if(in_deg[target]==0)
+        indeg0_nodes.push(target);
+    }
+  }
+
+  // if no edge is left in the graph, it is acyclic
+  bool is_dag=none_of(
+    in_deg.begin(),
+    in_deg.end(),
+    [](std::pair<node_indext, size_t> p) { return p.second>0; });
```
Comment (tautschnig) [inline src/util/graph.h]: Wouldn't `bool is_dag=nodelist.size()==nodes.size();` also do the trick?

**Comment by mgudemann** [comment]
@tautschnig thank you for the comments, I have adapted the PR

File: src/util/graph.h (hunk 6 above)
Comment (mgudemann) [inline src/util/graph.h]: ok, changed this, but then we assume `node_indext` to be an integral value

Comment (tautschnig) [inline src/util/graph.h]: I believe this is a reasonable (likely necessary) assumption.

File: src/util/graph.h [hunk 9]
```diff
@@ -570,6 +580,54 @@ void grapht<N>::make_chordal()
...
+      if(in_deg[target]==0)
+        indeg0_nodes.push(target);
+    }
+  }
+
+  // if all nodes are sorted, the graph is acyclic
+  bool is_dag=(nodelist.size()==nodes.size());
+
+  // return empty list in case of cyclic graph
+  if(!is_dag)
```
Comment (tautschnig) [inline src/util/graph.h]: Nit picking: you might now get rid of the `is_dag` temporary and use `if(nodelist.size()!=nodes.size())`

**Comment by mgudemann** [comment]
@tautschnig thanks, I have simplified the `bool is_dag` away

**Pull request merged**